process/block/metablock.go

```diff
@@ -1561,39 +1561,120 @@ func (mp *metaProcessor) ApplyBodyToHeader(hdr data.HeaderHandler, bodyHandler d
+		if err != nil {
+			return nil, err
+		}
+
+		finalHeader := block.EpochStartShardData{
+			ShardId:               lastNotarizedHdr.GetShardID(),
+			HeaderHash:            hdrHash,
+			RootHash:              lastNotarizedHdr.GetRootHash(),
+			FirstPendingMetaBlock: lastMetaHash,
+			LastFinishedMetaBlock: lastFinalizedMetaHash,
+		}
+
+		epochStart.LastFinalizedHeaders = append(epochStart.LastFinalizedHeaders, finalHeader)
+		lastNotarizedHeaders[i] = lastNotarizedHdr
+	}
+
+	pendingMiniBlocks, err := mp.pendingMiniBlocks.PendingMiniBlockHeaders(lastNotarizedHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, pendingMiniBlock := range pendingMiniBlocks {
+		recvShId := pendingMiniBlock.ReceiverShardID
+
+		currShardPendingMBs := epochStart.LastFinalizedHeaders[recvShId].PendingMiniBlockHeaders
+		currShardPendingMBs = append(currShardPendingMBs, pendingMiniBlock)
+	}
+
+	return epochStart, nil
+}
+
+func (mp *metaProcessor) getLastFinalizedMetaHashForShard(shardHdr *block.Header) ([]byte, []byte, error) {
+	var lastMetaHash []byte
+	var lastFinalizedMetaHash []byte
+
+	numAddedMetas := len(shardHdr.MetaBlockHashes)
+	if numAddedMetas > 0 {
+		lastMetaHash = shardHdr.MetaBlockHashes[numAddedMetas-1]
+	}
+
+	if numAddedMetas > 1 {
+		lastFinalizedMetaHash = shardHdr.MetaBlockHashes[numAddedMetas-2]
+		return lastMetaHash, lastFinalizedMetaHash, nil
+	}
+
+	for currentHdr := shardHdr; currentHdr.GetNonce() > 0 && currentHdr.GetEpoch() == shardHdr.GetEpoch(); {
+		oldShardHdr, err := process.GetShardHeader(currentHdr.GetPrevHash(), mp.dataPool.ShardHeaders(), mp.marshalizer, mp.store)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		if len(oldShardHdr.MetaBlockHashes) == 0 {
+			currentHdr = oldShardHdr
+			continue
+		}
+
+		numAddedMetas = len(oldShardHdr.MetaBlockHashes)
+		if numAddedMetas > 1 {
+			if len(lastMetaHash) == 0 {
+				lastMetaHash = oldShardHdr.MetaBlockHashes[numAddedMetas-1]
+				lastFinalizedMetaHash = oldShardHdr.MetaBlockHashes[numAddedMetas-2]
+				return lastMetaHash, lastFinalizedMetaHash, nil
+			}
+
+			if bytes.Equal(lastMetaHash, oldShardHdr.MetaBlockHashes[numAddedMetas-1]) {
+				lastFinalizedMetaHash = oldShardHdr.MetaBlockHashes[numAddedMetas-2]
+				return lastMetaHash, lastFinalizedMetaHash, nil
+			}
+
+			lastFinalizedMetaHash = oldShardHdr.MetaBlockHashes[numAddedMetas-1]
+			return lastMetaHash, lastFinalizedMetaHash, nil
+		}
+
+		if len(lastMetaHash) == 0 {
+			lastMetaHash = oldShardHdr.MetaBlockHashes[numAddedMetas-1]
+			currentHdr = oldShardHdr
+			continue
+		}
+
+		lastFinalizedMetaHash = oldShardHdr.MetaBlockHashes[numAddedMetas-1]
+		if !bytes.Equal(lastMetaHash, lastFinalizedMetaHash) {
+			return lastMetaHash, lastFinalizedMetaHash, nil
 		}
 
-		endOfEpoch.LastFinalizedHeaders = append(endOfEpoch.LastFinalizedHeaders, finalHeader)
+		currentHdr = oldShardHdr
 	}
 
-	return endOfEpoch, nil
+	return nil, nil, process.ErrLastFinalizedMetaHashForShardNotFound
 }
```

Comment:
in function getOrderedHdrs, instead of having some comment for code sections you can extract those into functions
e.g lines 1741 to 1772 -> getHeadersPerShard
lines 1776 to 1791 -> sortHeaders
lines 1794 to 1810 -> mapToList

Reply:
done